### ISSUE TITLE
twisted.python.components._ProxiedClassMethod could look more like a normal method object.

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#7832]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)| @exarkun reported|
|-|-|
|Trac ID|trac#7832|
|Type|enhancement|
|Created|2015-03-25 15:09:53Z|

`_ProxiedClassMethod` implements part of the standard ... "method protocol" ...  That is, it has a `__call__` method.  It's missing other things that a method typically has, though.

In particular, missing `__name__` makes it impossible to use `functools.wraps` with one of these proxied things.  The name information is readily available, it just gets put at the `methodName` attribute instead of the `__name__` attribute.

```
>>> from twisted.python.components import proxyForInterface
>>> from twisted.internet.interfaces import IReactorTCP
>>> from functools import wraps
>>> @wraps(proxyForInterface(IReactorTCP).connectTCP)
... def f():
...     pass
... 
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/usr/lib/python2.7/functools.py", line 33, in update_wrapper
    setattr(wrapper, attr, getattr(wrapped, attr))
AttributeError: '_ProxiedClassMethod' object has no attribute '__name__'
>>> 
```

Encountered while writing a class decorator which applied a decorator to one of the decorated class's methods.  This worked fine until I tried to decorate a class that used `proxyForInterface`.

Attachments:

* [7832.patch](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/bd5/bd5404782132a494f0da930e474780171293e760/bff527de82cf462c39f3b4fa9e298294941d5d99.patch) (2030 bytes) - added by exarkun on 2015-03-25 15:20:57Z - 
* [7832-wrong-docstring.patch](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/bd5/bd5404782132a494f0da930e474780171293e760/2655ec15f740b12c94376c774fa5a9924435c39c.patch) (3014 bytes) - added by herrwolfe on 2015-03-25 23:25:34Z - 
* [7832-2-wrong-docstring.patch](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/bd5/bd5404782132a494f0da930e474780171293e760/4880c71fc9da0b44da14b55b4f04d59d263e70c6.patch) (3007 bytes) - added by herrwolfe on 2015-03-25 23:34:07Z - exarkun's patch, failing doc test, fixed spacing, remove u"..."
* [7832-3-wrong-docstring.patch](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/bd5/bd5404782132a494f0da930e474780171293e760/73cc3588cce584e8c8db5d6c366352204d1adc9c.patch) (2328 bytes) - added by herrwolfe on 2015-03-25 23:43:22Z - fixes whitespace errors I introduced - sorry for spam.

<details><summary>Searchable metadata</summary>

```
trac-id__7832 7832
type__enhancement enhancement
reporter__exarkun exarkun
priority__normal normal
milestone__ 
branch__ 
branch_author__ 
status__closed closed
resolution__fixed fixed
component__core core
keywords__ 
time__1427296193734174 1427296193734174
changetime__1427383567866634 1427383567866634
version__None None
owner__exarkun exarkun

```
</details>

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)|@exarkun set owner to @exarkun|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)|@exarkun removed owner|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_3"></a>|@exarkun commented|
|-|-|

Development sponsored by ClusterHQ Inc.

**Comment by twisted-trac** [comment]
|[<img alt="wsanchez's avatar" src="https://avatars.githubusercontent.com/u/50002?s=50" width="50" height="50">](https://github.com/wsanchez)<a name="note_4"></a>|@wsanchez set owner to @exarkun|
|-|-|

Unclear whether the docstring should document `__name__`… I guess that's a known thing for methods to have, so that would be redundant?

Since the class `_ProxiedClassMethod` is private… are it's ivars considered public API?  If the `methodName` ivar is only used internally, renaming it rather than copying it seems sensible.

These two seem fuzzy to me, so I'll defer to exarkun on them.  As is, this change is sound, I think.

**Comment by twisted-trac** [comment]
|[<img alt="derwolfe's avatar" src="https://avatars.githubusercontent.com/u/641357?s=50" width="50" height="50">](https://github.com/derwolfe)<a name="note_5"></a>|@derwolfe commented|
|-|-|

exarkun, 

It looks like the ``` __doc__ ``` attribute is incorrect as well.

With your patch applied, the decorated method can indeed be wrapped and called. However, when that method's f.__doc__ attribute is accessed the docstring for ``` t.w.c.proxyForInterface ``` is returned.

I've attached a patch, https://twistedmatrix.com/trac/attachment/ticket/7832/7832-3-wrong-docstring.patch, containing your changes and a failing test showing that the docstring returned does not match the docstring of the function being wrapped.

Thanks!

(sorry for spamming the ticket with multiple patches - I introduced whitespace errors that I did not notice until it was too late).

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_6"></a>|@exarkun commented|
|-|-|

Thanks for the additional fixes herrwolfe.  I'm reluctant to include them here as I don't personally have a requirement for the `__doc__` behavior right now.  It does seem like it is probably an improvement but perhaps it can be handled as part of a separate ticket.  (Also, I notice the new test method name doesn't follow the Twisted naming convention.)

> Unclear whether the docstring should document __name__… I guess that's a known thing for methods to have, so that would be redundant?

Probably doesn't hurt to be explicit.  I'll add it.

> Since the class _ProxiedClassMethod is private… are it's ivars considered public API? If the methodName ivar is only used internally, renaming it rather than copying it seems sensible.

They are public unfortunately because you can reach them without ever using a private API.  For example, `proxyForInterface(IFoo)(foo).foo_method.methodName`.  So we'll have to stick with being redundant.

I'll make documentation change and then apply this to trunk.  Thanks!

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_7"></a>|@exarkun set status to `closed`|
|-|-|

(In [44292]) Apply 7832.patch with `__name__` documentation added.

Author: exarkun
Reviewer: wsanchez
Fixes: [#7832](https://github.com/twisted/twisted/issues/7832)

Give proxied methods created by `proxyForInterface` a `__name__` attribute as a normal
method would have.
